aten/src/ATen/native/cuda/Normalization.cu

```diff
@@ -1,72 +1,278 @@
+#include <ATen/native/TensorIterator.h>
+#include <ATen/native/ReduceOps.h>
+#include <ATen/native/Resize.h>
+#include <ATen/native/cuda/Loops.cuh>
+#include <ATen/native/cuda/Reduce.cuh>
 #include <ATen/native/cuda/Normalization.cuh>
+#include <c10/cuda/CUDAMathCompat.h>
+
+namespace at { namespace native {
+
+namespace {
 
 inline bool batch_norm_use_channels_last_kernels(const at::Tensor& self) {
   return self.is_contiguous(at::MemoryFormat::ChannelsLast) || self.ndimension() == 2;
 }
 
-namespace at { namespace native {
+enum class Impl {
+  Contiguous,
+  ChannelsLast,
+  General,
+};
+
+inline Impl batch_norm_choose_impl(const Tensor& self) {
+  if (!at::cuda::detail::canUse32BitIndexMath(self)) {
+    return Impl::General;
+  }
+
+  if (self.is_contiguous()) {
+    return self.dim() == 2 ? Impl::ChannelsLast : Impl::Contiguous;
+  }
+
+  if (self.is_contiguous(at::MemoryFormat::ChannelsLast)) {
+    return Impl::ChannelsLast;
+  }
+
+  return Impl::General;
+}
+
+void batch_norm_elementwise(
+    const Tensor& out, const Tensor& self, const c10::optional<Tensor>& weight_opt,
+    const c10::optional<Tensor>& bias_opt, const Tensor& mean_, const Tensor& invstd_) {
+  const double dummy_epsilon = 1e-5;
```

Comment:
can you please leave a comment here that batch_norm_elemt_cuda_template doesn't use epsilon for training?

Reply:
I've added a comments for the two `dummy_epsilons` and moved this one into the template where train mode is set.